Handle Doze whitelisting after onboarding gracefully.

denzilferreira opened the pull request:
This pull request resolves #31 

This is handled by Google Play Services and there is no manual toggling of Doze settings, which are confusing to majority of users.

denzilferreira committed:
Merge pull request #1 from THLfi/trunk

Update to latest changes

denzilferreira committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

denzilferreira committed:
Fix #31
- Handle Doze whitelisting automatically after onboarding.

denzilferreira committed:
Fix #31
- Handle Doze whitelisting automatically after onboarding.
- Updated CHANGELOG.md

denzilferreira committed:
Merge pull request #2 from THLfi/trunk

Run owasp release scan against latest github action definitions (#37)

solita-jussihei commented:
Hi @denzilferreira and thanks for the pull request! This is a good suggestion.. couple of questions:
- It seems that it's ok to fire the intent every time app starts, and it won't do anything if user already accepted? But it will repeat the prompt, if user denied last time?
- We might want to do this check in onResume instead of onCreate, so that we could catch if user changes settings and returns to the app?
- Maybe also add a check to PowerManager.isIgnoringBatteryOptimizations() to see if already ignoring optimizations?

denzilferreira committed:
Latest changes

denzilferreira committed:
Merge branch 'THLfi-trunk' into trunk

denzilferreira commented:
@solita-jussihei

[x] Maybe also add a check to PowerManager.isIgnoringBatteryOptimizations() to see if already ignoring optimizations?
 
> It seems that it's ok to fire the intent every time app starts, and it won't do anything if user already accepted? But it will repeat the prompt, if user denied last time?
> We might want to do this check in onResume instead of onCreate, so that we could catch if user changes settings and returns to the app?

This is one of the reasons I had it inside onCreate: as it's only executed once. If onResume, every time we press deny, we come back to onResume and ask again to let the app run on the background. Perhaps a shared preference to store the user's preference and not ask again if they have denied it?

denzilferreira committed:
- Add two use cases for whitelisting: manual and automated in checkDoze(manualToggle)

TODO:
- Store preference for not using doze
- GUI for instructions for manual doze toggle
- GUI for rationale for automated doze whitelisting?

denzilferreira committed:
- Use a state machine to track doze whitelisting state: denied, allowed, unknown
- Persist the user's whitelisting preference

denzilferreira committed:
Merge branch 'trunk' into trunk

denzilferreira committed:
- Use isWhitelisted(): Boolean to check the current status of Doze
- Update doze state machine if user manually changes battery optimization

solita-jussihei commented on app/src/main/java/fi/thl/koronahaavi/MainActivity.kt:
```diff
@@ -121,6 +141,16 @@ class MainActivity : AppCompatActivity() {
         super.onActivityResult(requestCode, resultCode, data)
         resolutionViewModel.handleActivityResult(requestCode, resultCode)
+
+        resolutionViewModel.whitelistResolvedEvent().observe(this, Observer {
```
Hi Denzil, this livedata observation should probably be setup in onCreate.. but we will work on that when integrating these changes, thanks for the work on this!
